Tests/BSONTests/ExtendedJSONConversionTests.swift

```diff
@@ -0,0 +1,27 @@
+@testable import BSON
+import Foundation
+import Nimble
+import NIO
+import XCTest
+
+open class ExtendedJSONConversionTestCase: XCTestCase {
+    func testInt32() throws {
+        // Success cases
+        expect(try Int32(fromExtJSON: 5)).to(equal(5))
+        expect(try Int32(fromExtJSON: ["$numberInt": "5"])).to(equal(5))
+
+        // Nil cases
+        expect(try Int32(fromExtJSON: JSON.number(Double(Int32.max) + 1))).to(beNil())
+        expect(try Int32(fromExtJSON: JSON.bool(true))).to(beNil())
+
+        // Error cases
+        expect(try Int32(fromExtJSON: ["$numberInt": "5", "extra": true]))
+            .to(throwError(errorType: BSONError.InternalError.self))
+        expect(try Int32(fromExtJSON: ["$numberInt": .number(Double(Int32.max) + 1)]))
+            .to(throwError(errorType: BSONError.InternalError.self))
+        expect(try Int32(fromExtJSON: ["bad": "5"]))
```

Comment:
this should be a nil case, since it just doesnt match up with the extended JSON format of an `Int32`. The other two are errors because they indicate that an `Int32` is being specified, but then provide values that are incompatible with the type (e.g. overflowing `Int32`) or specify an invalid format (an extra field).